### PR TITLE
Eink weact 2.13/2.9 support

### PR DESCRIPTION
**Pull request by alinelena**
# What does this implement/fix?

<!-- Quick description and explanation of changes -->

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>
https://github.com/esphome/feature-requests/issues/2101

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
- platform: waveshare_epaper                                                                                                                                                                                                                        
  model: "2.90inBS" 
  model: "2.13inBN" 

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by lboue** [comment]
Hello. Can you merge this PR?

File: esphome/components/waveshare_epaper/display.py [hunk 1]
```diff
@@ -142,7 +148,10 @@ async def to_code(config):
     if model_type == "a":
         rhs = WaveshareEPaperTypeA.new(model)
         var = cg.Pvariable(config[CONF_ID], rhs, WaveshareEPaperTypeA)
-    elif model_type in ("b", "c"):
+    elif model_type == "c":
+        rhs = WaveshareEPaperTypeC.new(model)
+        var = cg.Pvariable(config[CONF_ID], rhs, WaveshareEPaperTypeC)
+    elif model_type in ("b"):
```
Comment (jesserockz) [inline esphome/components/waveshare_epaper/display.py]: This is going to change/break for any existing 'c' type displays?

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by kungjohan** [comment]
As I have described in https://github.com/esphome/feature-requests/issues/2191 I have problems with running this on ESPhome later than 2023.12.0.

Is it possible to update this pull request to work with acutal ESPHome version?

**Comment by alinelena** [comment]
> As I have described in [esphome/feature-requests#2191](https://github.com/esphome/feature-requests/issues/2191) I have problems with running this on ESPhome later than 2023.12.0.
> 
> Is it possible to update this pull request to work with acutal ESPHome version?

probably yes... I will try to find some time this weekend. I will also need to see if upstream did something useful

**Comment by kungjohan** [comment]
I tried to port your code on upstream dev. I have only tested it on black and white 2.9 WeAct module. 
My efforts are located at:
https://github.com/kungjohan/esphome/tree/weact_2_9

File: esphome/components/waveshare_epaper/display.py (hunk 1 above)
Comment (alinelena) [reply esphome/components/waveshare_epaper/display.py]: no idea, but things changed so much that probably all this needs refactored.

**Comment by alinelena** [comment]
2.13 works not on dev with 2.13inv3 which was recently added. as mentioned before this will need to be refactored to be aligned with new changes in the component

**Comment by alinelena** [comment]
@kungjohan had a look so now 2.13 is supported by dev by using the new 2.13inv3 for 2.9 one you need some changes but are minimal... so easier to maintain till one gets best way to get it merged

```c
diff --git a/esphome/components/waveshare_epaper/waveshare_213v3.cpp b/esphome/components/waveshare_epaper/waveshare_213v3.cpp
index 196aeed3..8e8f50e8 100644
--- a/esphome/components/waveshare_epaper/waveshare_213v3.cpp
+++ b/esphome/components/waveshare_epaper/waveshare_213v3.cpp
@@ -52,9 +52,9 @@ static const uint8_t BORDER_PART[] = {0x3C, 0x80};  // border waveform
 static const uint8_t BORDER_FULL[] = {0x3C, 0x05};  // border waveform
 static const uint8_t CMD1[] = {0x3F, 0x22};
 static const uint8_t RAM_X_START[] = {0x44, 0x00, 121 / 8};           // set ram_x_address_start_end
-static const uint8_t RAM_Y_START[] = {0x45, 0x00, 0x00, 250 - 1, 0};  // set ram_y_address_start_end
+static const uint8_t RAM_Y_START[] = {0x45, 0x00, 0x00, uint8_t(296-1),(296 - 1)>>8};  // set ram_y_address_start_end
 static const uint8_t RAM_X_POS[] = {0x4E, 0x00};                      // set ram_x_address_counter
-// static const uint8_t RAM_Y_POS[] = {0x4F, 0x00, 0x00};        // set ram_y_address_counter
+static const uint8_t RAM_Y_POS[] = {0x4F, 0x00, 0x00};        // set ram_y_address_counter
 #define SEND(x) this->cmd_data(x, sizeof(x))
 
 void WaveshareEPaper2P13InV3::write_lut_(const uint8_t *lut) {
@@ -164,7 +164,7 @@ void WaveshareEPaper2P13InV3::display() {
 
 int WaveshareEPaper2P13InV3::get_width_internal() { return 128; }
 
-int WaveshareEPaper2P13InV3::get_height_internal() { return 250; }
+int WaveshareEPaper2P13InV3::get_height_internal() { return 296; }
 
 uint32_t WaveshareEPaper2P13InV3::idle_timeout_() { return 5000; }
 ```

once i get an idea of the most acceptable way to add 2.9 support I create a new pr and superseed this one.

**Comment by kungjohan** [comment]
I can not see that RAM_Y_POS is used is that a typo?

What do you think of (not tested, but as an idea) and also adding corresponding to display.py:
```
diff --git a/esphome/components/waveshare_epaper/waveshare_epaper.h b/esphome/components/waveshare_epaper/waveshare_epaper.h
index 0f1144cc..d5c70c39 100644
--- a/esphome/components/waveshare_epaper/waveshare_epaper.h
+++ b/esphome/components/waveshare_epaper/waveshare_epaper.h
@@ -639,7 +639,7 @@ class WaveshareEPaper2P13InV3 : public WaveshareEPaper {
   uint32_t idle_timeout_() override;
 
   void write_buffer_(uint8_t cmd, int top, int bottom);
-  void set_window_(int t, int b);
+  virtual void set_window_(int t, int b);
   void send_reset_();
   void partial_update_();
   void full_update_();
@@ -649,5 +649,26 @@ class WaveshareEPaper2P13InV3 : public WaveshareEPaper {
   bool is_busy_{false};
   void write_lut_(const uint8_t *lut);
 };
+
+class WaveshareEPaper2P9InWeAct : public WaveshareEPaper2P13InV3 {
+ protected:
+  static const uint8_t RAM_Y_START[] = {0x45, 0x00, 0x00, uint8_t(296-1),(296 - 1)>>8};  
+  static const uint8_t RAM_Y_POS[] = {0x4F, 0x00, 0x00};
+
+  int get_height_internal() override { return 296; }
+
+  void set_window_(int t, int b) override {
+  uint8_t buffer[3];
+
+  SEND(RAM_X_START);
+  SEND(RAM_Y_START);
+  SEND(RAM_X_POS);
+  buffer[0] = 0x4F;
+  buffer[1] = (uint8_t) t;
+  buffer[2] = (uint8_t) (t >> 8);
+  SEND(buffer);
+}
+  
+};
 }  // namespace waveshare_epaper
 }  // namespace esphome

```

**Comment by jbergler** [comment]
Hello, I stumbled across this PR while trying to get the 2.9" 3 color (black, white, red) module working.
If I understand the 3 color displays, the update time is much higher, and this results in the a timeout updating the display.

I've had to make a few changes:
- Call `App.feed_wdt();` in `wait_until_idle` to prevent the esp watchdog from resetting the display.
- Increase the timeout to 30s (updates appear to be completing in ~21.5s)

**Comment by alinelena** [comment]
> @kungjohan had a look so now 2.13 is supported by dev by using the new 2.13inv3 for 2.9 one you need some changes but are minimal... so easier to maintain till one gets best way to get it merged
> 
> ```c
> diff --git a/esphome/components/waveshare_epaper/waveshare_213v3.cpp b/esphome/components/waveshare_epaper/waveshare_213v3.cpp
> index 196aeed3..8e8f50e8 100644
> --- a/esphome/components/waveshare_epaper/waveshare_213v3.cpp
> +++ b/esphome/components/waveshare_epaper/waveshare_213v3.cpp
> @@ -52,9 +52,9 @@ static const uint8_t BORDER_PART[] = {0x3C, 0x80};  // border waveform
>  static const uint8_t BORDER_FULL[] = {0x3C, 0x05};  // border waveform
>  static const uint8_t CMD1[] = {0x3F, 0x22};
>  static const uint8_t RAM_X_START[] = {0x44, 0x00, 121 / 8};           // set ram_x_address_start_end
> -static const uint8_t RAM_Y_START[] = {0x45, 0x00, 0x00, 250 - 1, 0};  // set ram_y_address_start_end
> +static const uint8_t RAM_Y_START[] = {0x45, 0x00, 0x00, uint8_t(296-1),(296 - 1)>>8};  // set ram_y_address_start_end
>  static const uint8_t RAM_X_POS[] = {0x4E, 0x00};                      // set ram_x_address_counter
> -// static const uint8_t RAM_Y_POS[] = {0x4F, 0x00, 0x00};        // set ram_y_address_counter
> +static const uint8_t RAM_Y_POS[] = {0x4F, 0x00, 0x00};        // set ram_y_address_counter
>  #define SEND(x) this->cmd_data(x, sizeof(x))
>  
>  void WaveshareEPaper2P13InV3::write_lut_(const uint8_t *lut) {
> @@ -164,7 +164,7 @@ void WaveshareEPaper2P13InV3::display() {
>  
>  int WaveshareEPaper2P13InV3::get_width_internal() { return 128; }
>  
> -int WaveshareEPaper2P13InV3::get_height_internal() { return 250; }
> +int WaveshareEPaper2P13InV3::get_height_internal() { return 296; }
>  
>  uint32_t WaveshareEPaper2P13InV3::idle_timeout_() { return 5000; }
> ```
> 
> once i get an idea of the most acceptable way to add 2.9 support I create a new pr and superseed this one.

yes no idea why the hell I left the line un-commented. 

> I can not see that RAM_Y_POS is used is that a typo?
> 
> What do you think of (not tested, but as an idea) and also adding corresponding to display.py:
> 
> ```
> diff --git a/esphome/components/waveshare_epaper/waveshare_epaper.h b/esphome/components/waveshare_epaper/waveshare_epaper.h
> index 0f1144cc..d5c70c39 100644
> --- a/esphome/components/waveshare_epaper/waveshare_epaper.h
> +++ b/esphome/components/waveshare_epaper/waveshare_epaper.h
> @@ -639,7 +639,7 @@ class WaveshareEPaper2P13InV3 : public WaveshareEPaper {
>    uint32_t idle_timeout_() override;
>  
>    void write_buffer_(uint8_t cmd, int top, int bottom);
> -  void set_window_(int t, int b);
> +  virtual void set_window_(int t, int b);
>    void send_reset_();
>    void partial_update_();
>    void full_update_();
> @@ -649,5 +649,26 @@ class WaveshareEPaper2P13InV3 : public WaveshareEPaper {
>    bool is_busy_{false};
>    void write_lut_(const uint8_t *lut);
>  };
> +
> +class WaveshareEPaper2P9InWeAct : public WaveshareEPaper2P13InV3 {
> + protected:
> +  static const uint8_t RAM_Y_START[] = {0x45, 0x00, 0x00, uint8_t(296-1),(296 - 1)>>8};  
> +  static const uint8_t RAM_Y_POS[] = {0x4F, 0x00, 0x00};
> +
> +  int get_height_internal() override { return 296; }
> +
> +  void set_window_(int t, int b) override {
> +  uint8_t buffer[3];
> +
> +  SEND(RAM_X_START);
> +  SEND(RAM_Y_START);
> +  SEND(RAM_X_POS);
> +  buffer[0] = 0x4F;
> +  buffer[1] = (uint8_t) t;
> +  buffer[2] = (uint8_t) (t >> 8);
> +  SEND(buffer);
> +}
> +  
> +};
>  }  // namespace waveshare_epaper
>  }  // namespace esphome
> ```

no idea why i left the line un-commented. since the changes are so trivial I wanted a comment from esphome devs to see what is acceptable for them to get it in.

File: esphome/components/waveshare_epaper/display.py (hunk 1 above)
Comment (alinelena) [reply esphome/components/waveshare_epaper/display.py]: so this become irrelevant in the view of latest changes in the esphome... changes now to support 2.9inch which is not supported are minimial, see comment https://github.com/esphome/esphome/pull/4579#issuecomment-1908812736 and below. still will need advice on best way to get things in.

**Comment by alinelena** [comment]
sorry I closed this by mistake, when trying to rebase to add the changes... opened a new one #6217

**Pull request closed without merging**